src/rgw/rgw_amqp.cc

```diff
@@ -129,111 +225,151 @@ amqp_connection_state_t create_connection(const amqp_connection_info& info, cons
       info.user,
       info.password);
   if (reply.reply_type != AMQP_RESPONSE_NORMAL) {
-    // TODO: report the specific error and retry
-    return nullptr;
+    throw(connection_error("failed to login to broker: " + to_string(reply)));
+    return NO_CONNECTION;
   }
 
   // open channel
-  amqp_channel_open(conn, CHANNEL_ID);
-  if (amqp_get_rpc_reply(conn).reply_type != AMQP_RESPONSE_NORMAL) {
-    // TODO: report the specific error and retry
-    return nullptr;
+  {
+    const auto ok = amqp_channel_open(conn, CHANNEL_ID);
+    if (!ok) {
+      throw(connection_error("failed to open channel: " + std::to_string(CHANNEL_ID)));
+      return NO_CONNECTION;
+    }
   }
 
   // verify that the topic exchange is there
-  amqp_exchange_declare(conn, 
+  {
+    const auto ok = amqp_exchange_declare(conn, 
       CHANNEL_ID,
       amqp_cstring_bytes(exchange.c_str()),
       amqp_cstring_bytes("topic"),
-      0, // TODO: should be passive - exchange must already exist on broker
+      1, // passive - exchange must already exist on broker
       1, // durable
       0, // dont auto-delete
       0, // not internal
       amqp_empty_table);
-
-  if (amqp_get_rpc_reply(conn).reply_type != AMQP_RESPONSE_NORMAL) {
-    // TODO: report the specific error and retry
-    return nullptr;
+    if (!ok) {
+      throw(connection_error("exchange not found"));
+      return NO_CONNECTION;
+    }
+    const auto reply = amqp_get_rpc_reply(conn);
+    if (reply.reply_type != AMQP_RESPONSE_NORMAL) {
+      throw(connection_error("failed to declare exchange: " + to_string(reply)));
+      return NO_CONNECTION;
+      // TODO: retry 
+    }
   }
 
   guard.reset();
-  return conn;
+  return connection_t(conn, exchange);
 }
 
-typedef std::map<amqp_connection_info, amqp_connection_state_t> ConnectionList;
+typedef std::unordered_map<ConnectionID, connection_t, ConnectionID::hasher> ConnectionList;
 
-class Manager {
+class Manager : public std::thread {
```

Comment:
is it common to derive from std::thread?

Reply:
Not sure what are the cons/pros here. In general I prefer composition to inheritance, so, I'm ok with changing.